matminer/featurizers/structure.py

```diff
@@ -1590,6 +1594,182 @@ def citations(self):
+
+    def fit(self, X, y=None):
+        """
+        Define the bags using a list of structures.
+
+        Both the names of the bags (e.g., Cs-Cl) and the maximum lengths of
+        the bags are set with fit.
+
+        Args:
+            X (Series/list): An iterable of pymatgen Structure
+                objects which will be used to determine the allowed bond
+                types and bag lengths.
+            y : unused (added for consistency with overridden method signature)
+
+        Returns:
+            self
+
+        """
+        unpadded_bobs = [self.bag(s) for s in X]
+        bonds = [list(bob.keys()) for bob in unpadded_bobs]
+        bonds = np.unique(sum(bonds, []))
+        baglens = [0]*len(bonds)
+
+        for i, bond in enumerate(bonds):
+            for bob in unpadded_bobs:
+                if bond in bob:
+                    baglen = len(bob[bond])
+                    baglens[i] = max((baglens[i], baglen))
+        self.baglens = dict(zip(bonds, baglens))
+        # Sort the bags by bag length, with the shortest coming first.
+        self.ordered_bonds = [b[0] for b in sorted(self.baglens.items(),
+                                    key=lambda bl: bl[1])]
+
+    def bag(self, s):
+        """
+        Convert a structure into a bag of bonds, where each bag has no padded
+        zeros. using this function will give the 'raw' bags, which when
+        concatenated, will have different lengths.
+
+        Args:
+            s (Structure): A pymatgen Structure or IStructure object.
+
+        Returns:
+            (dict) A bag of bonds, where the keys are sorted tuples of pymatgen
+                Site objects representing bonds or sites, and the values are the
+                Coulomb matrix values for that bag.
+        """
+        if not isinstance(s, (Structure)):
+            raise TypeError("BagofBonds must take in a pymatgen Structure "
+                            "object, not {}".format(type(s)))
+
+        cm = self.coulomb_matrix.featurize(s)[0]
+        sites = s.sites
+        nsites = len(sites)
+        bonds = np.zeros((nsites, nsites), dtype=object)
+        for i, si in enumerate(sites):
+            for j, sj in enumerate(sites):
+                el0, el1 = si.specie, sj.specie
+                if isinstance(el0, Specie):
+                    el0 = el0.element
+                if isinstance(el1, Specie):
+                    el1 = el1.element
+                if i == j:
+                    bonds[i, j] = (el0,)
+                else:
+                    bonds[i, j] = tuple(sorted((el0, el1)))
+        bags = {b: [] for b in np.unique(bonds)}
+        for i in range(nsites):
+            for j in range(nsites):
+                bond = bonds[i, j]
+                cmval = cm[i, j]
+                bags[bond].append(cmval)
+
+        # We must sort the magnitude of bonds in each bag
+        return {bond: sorted(bags[bond]) for bond in bags}
+
+
+    def featurize(self, s):
```

Comment:
Can you add a docstring here like you did for bag()?

Some of the docstrings in matminer are inconsistent but we'll at least try to be more consistent for new featurizers and fix the old ones later

Reply:
For featurize? Sure.

Reply:
yes for featurize